### ISSUE TITLE
.gitignore not ignoring files and directories that start with .

### ISSUE THREAD

**Issue by adam-lync** [git-for-windows/git#2804]
.gitignore not ignoring files and directories that start with .

## Description
In VS Code Remote-containers extension requires .devcontainer directory or .devcontainer.json file to create remote-container. .gitignore file is being used to stop unintended changes to this file/directory because people are error prone and I don't need 500 developers ticked off because their remote-containers have stopped working.

## Expected Behavior
.gitignore would ignore any changes to files in .devcontainer directory or the .devcontainer.json file when the following are entered into the .gitignore file

.devcontainer/
.devcontainer.json

## Actual Behavior
.gitignore does not ignore these files. Changing directory or file name to devcontainer or devcontainer.json and updating .gitignore works indicating it is an issue with the beginning dot.

## Possible Fix
git is not correctly identifying the dot at the beginning of the directory or filename or ignoring its ignore file when it sees these items.

## Steps to Reproduce
1. Create file .devcontainer.json or directory .devcontainer
2. Create .gitignore and added file or directory name to file
3. After committing and clearing git cache, make change to .devcontainer.json or .devcontainer directory
4. Git attempts to commit changes

## Your Environment
* Version used: git version 2.28.0.windows.1
* Environment name and version (e.g. Chrome 39, node.js 5.4): VS Code 1.48.2
* Operating System and version (desktop or mobile): Windows 10 1909
* Link to your project: Private Bitbucket repo

**Comment by dscho** [comment]
I tried to reproduce this issue, but failed:

```
me@work MINGW64 ~/repros/gitignore-devcontainer-2804 (master)
$ >.devcontainer.json

me@work MINGW64 ~/repros/gitignore-devcontainer-2804 (master)
$ echo .devcontainer.json >.gitignore

me@work MINGW64 ~/repros/gitignore-devcontainer-2804 (master)
$ git status
On branch master

No commits yet

Untracked files:
  (use "git add <file>..." to include in what will be committed)
        .gitignore

nothing added to commit but untracked files present (use "git add" to track)

```

As you see, the `.devcontainer.json` file does _not_ show up.

However, I have a hunch that I know what is actually going on on your side (pro tip: next time, provide a [Minimal, Complete & Verifiable Example](https://stackoverflow.com/help/mcve), then nobody has to guess): you probably _already_ staged, or even committed, the `.devcontainer.json` file. In that case, Git has received conflicting information: on the one hand, it was explicitly asked to track the file (by staging it), on the other hand it was asked to ignore it (by adding it to `.gitignore`). Since those two instructions are contradicting each other, Git picks one side: the one where it tracks the file.

**Comment by adam-lync** [comment]
@dscho I had already committed .devcontainer.json to add it to the repo so that any clones would include it in VS Code but I didn't want anyone to be able to commit new changes hence the inclusion of .gitignore. I attempted the fix here: [https://stackoverflow.com/questions/1139762/ignore-files-that-have-already-been-committed-to-a-git-repository](https://stackoverflow.com/questions/1139762/ignore-files-that-have-already-been-committed-to-a-git-repository) which would remove the .devcontainer.json file from the repo which I didn't want. That is what led to this being created as I couldn't find a way to make this work at the time. I just came across `git update-index --assume-unchanged` which works perfectly how I need it to.

Thanks for the help.
